EIPS/eip-6617.md

```diff
@@ -14,18 +14,122 @@ created: 2023-02-27
 
 This EIP offers a standard for building a bit-based permission and role system. Each permission is represented by a single bit. By using an `uint256`, up to $256$ permissions and $2^{256}$ roles can be defined. We are able to specify the importance of each permission based on the order of the bits.
 
-
 ## Specification
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
 
-*Note* The following specifications use syntax from Solidity `0.8.7` (or above)
+_Note_ The following specifications use syntax from Solidity `0.8.7` (or above)
 
 - Permission and role MUST be defined as an `uint256`
```

Comment:
I appreciate author's effort to propose and author this EIP, folks. Hope this help explain it better about what kind of improvement I personally would like to see in this EIP before it can be merged as a Review

The specification of EIP at this current snapshot is far from editorially ready.

For example, It begins with  "- Permission MUST be defined as a power of two"

This is confusing. Is it a requirement for a complying contract to "have the methods of Permission" ? If so, it shall establish a requirement for certain methods. Fo example

A compliant contract MUST have the following interface...

And then, it could talk about the requirement of each method, such as "Permission MUST be defined as a power of two."

Also, what does "Permission MUST be defined as power of two" means, exactly?
Maybe I think you mean that Permission is bit-wise so each Permission is defined as one bit in a bytes32/uint256.
etc. 

Then the implementation shall refer to these requirements shows how a typical example of permission is implemented and how each methods will interact with it.


@SamWilsn FYI: "I think my last complaint before merging as draft is that this should be an interface.".

Reply:
Thanks for the comments @xinbenlv,
I tried to apply all the requests everyone made, tell me if it's better like that :smile: